### DISCUSSION TITLE
Extend documentation for patterns

### DISCUSSION THREAD

**Discussion by Phantasus** [lfe/lfe#478] [Q&A] [answered]
Hi,

please extend the documentation so that for certain syntaxes it's more clear to write
patterns in that syntax. The documentation reads rather academically when compared
to Elixir and it makes it hard to understand.

For example currently I struggle with to find out of how
to write patterns for lambda expressions. I know that technically this is something
Erlang and Elixir are capable of as I have written such patterns.

But under lfe, I have no clue, even after consulting the guide and the documentation
of how I'm supposed to write that.

```
lfe> (funcall (lambda ((tuple 1 2)) 'yes) (tuple 1 2))
** exception error: bad lambda form
```

**Comment by rvirding** [comment, marked as answer]
@Phantasus sorry I thought I had answered this, hence the label, but I must have forgotten to save it.

So the standard `lambda` which you were experimenting with does not handle patterns in the arguments. It is very CL basic and just takes a list of argument symbols/variables. When you want to have pattern matching and guard tests then you need to use `match-lambda` which has multiple **clauses** which it can test to choose which one. The basic syntax is
```
(match-lambda
  (( arg-patterns ) (when tests) body-expressions)
  (( arg-patterns ) (when tests) body-expresssion)
  ...)
```
Note that the `(when  tests)` is optional. So you example would be written as
```
lfe> (funcall (match-lambda (((tuple 1 2)) 'yes)) (tuple 1 2))
yes
```
If we want mulitple clauses then an example would be
```
lfe> (funcall (match-lambda ([(tuple 1 2)] 'yes)
lfe>                        ([(tuple x y)] (+ x y))) (tuple 99 100))
199
```
Note I am using `[ ... ]` around the arguments instead of `( ... )`. They are completely equivalent and I have just chosen to use them to make it easier to read. I could in fact use them everywhere. :smile:

NOTE: that each clause must have the same number of arguments as a function takes a specific number of arguments.

**Comment by rvirding** [comment]
Final comment I have the same options when defining a function with `defun` so the functions definitions could look like
```
(defun my-func
   ([(tuple 1 2)] 'yes)
   ([(tuple x y)] (+ x y)))
```

**Comment by oubiwann** [comment]
@Phantasus Just a heads-up on on the reference documents in the LFE repo: they are more like language specs (a la BNF) than actual documentation. 

But! Good news 😄 -- I have created a couple of tickets for projects where we're working on more extensive LFE documentation:
* https://github.com/cnbbooks/lfe-manual/issues/27
* https://github.com/cnbbooks/lfe-tutorial/issues/23
